[bitnami/keycloak] fix configMap cert directory permission issues

jgallucci32 opened the pull request:
Signed-off-by: jgallucci32 <john.gallucci.iv@gmail.com>


<!--
 Before you open the request please review the following guidelines and tips to help it be more easily integrated:

 - Describe the scope of your change - i.e. what the change does.
 - Describe any known limitations with your change.
 - Please run any tests or examples that can exercise your modified code.

 Thank you for contributing! We will try to test and integrate the change as soon as we can, but be aware we have many GitHub repositories to manage and can't immediately respond to every request. There is no need to bump or check in on a pull request (it will clutter the discussion of the request).

 Also don't be worried if the request is closed or not integrated sometimes the priorities of Bitnami might not match the priorities of the pull request. Don't fret, the open source community thrives on forks and GitHub makes it easy to keep your changes in a forked repo.
 -->

**Description of the change**
Fixes an issue with the security context of the container image not allowing certificates to be copied to the `/opt/bitnami/keycloak/certs` directory. Also fixes a mismatch in the configMap for the environment variables to use the correct cert directory (previously was using a non-existent directory). Also simplifies the deployment to use a single TLS certificate rather than having to create one per replica.
<!-- Describe the scope of your change - i.e. what the change does. -->

**Benefits**
Properly configures Keycloak to use a custom TLS certificate.
<!-- What benefits will be realized by the code change? -->

**Possible drawbacks**
Simplifying TLS certificate to a single cert changes the previous method of using unique certificates for each replica of a stateful set. This is an edge case and is unknown if it is actually used in Kubernetes deployments. Given this chart did not function when initially created 2 months ago, the risk is low.
<!-- Describe any known limitations with your change -->

**Applicable issues**

<!-- Enter any applicable Issues here (You can reference an issue using #) -->
  - fixes #4821

**Additional information**

<!-- If there's anything else that's important and relevant to your pull
request, mention that information here.-->

**Checklist** <!-- [Place an '[X]' (no spaces) in all applicable fields. Please remove unrelated fields.] -->
- [ ] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [ ] Variables are documented in the README.md
- [ ] Title of the PR starts with chart name (e.g. `[bitnami/chart]`)
- [ ] If the chart contains a `values-production.yaml` apart from `values.yaml`, ensure that you implement the changes in both files

:warning: Keep in mind that if you want to make changes to the kubeapps chart, please implement them in the [kubeapps repository](https://github.com/kubeapps/kubeapps/tree/master/chart/kubeapps). This is only a synchronized mirror.

jgallucci32 commented:
/hold

While the chart does fix the cert directory issue, the underlying problem of the container using the TLS settings correctly does not get fixed until version `11.0.3-debian-10-r63` of the container. The current version of the container in the chart is `11.0.3-debian-10-r59`. So do I need to bump the container version to `r63` or a more current version to fully fix TLS settings?

FraPazGal commented:
Hello @jgallucci32,

Looking into this, I think the option for multiple keystores is to ensure that internal cluster secure communication is possible. I think we can readily solve the permissions error when creating the directory with an initContainer and a mounted volume. We could transfer the code done in the container's command to the initContainer where we won't have any permissions-related error, what do you think?

Regarding the container's version used in the chart, our internal tools will update them before merging the PR, so you do not need to change it.

jgallucci32 commented:
@FraPazGal Thanks for note about chart version. I created a separate PR to remove it from the checklist since it's handled by the internal tools https://github.com/bitnami/charts/pull/5014

If all replicas used the same certificate, shouldn't internal TLS communication still function normally? How do we validate this?

If we go down this path 3 volume mounts will be requied. The initContainer will need to mount the configMap and a second shared volume between the primary container.

FraPazGal commented:
Hello @jgallucci32,

I see my colleague responded to your other PR that the checklist still has value, but thanks for taking the time anyways!

> If all replicas used the same certificate, shouldn't internal TLS communication still function normally? How do we validate this?

I don't think that would be the case as each pod would have a different hostname. I''m not sure if with the default configuration the pods communicate with each other, but they would do so in a cluster configuration. I'll prefer to use 3 volumes like you said; we would solve the current issue without dropping any functionalities and keep the door open for future additions.

jgallucci32 commented:
@FraPazGal I updated the PR to use an initContainer to configure the certs with a shared volume which is mounted to both containers. I also updated the logic of obtaining the ID of the replica to use a Kubernetes variable rather than rely on the Helm chart to inject it during the initial deployment.

FraPazGal commented on bitnami/keycloak/templates/statefulset.yaml:
```diff
@@ -59,9 +59,43 @@ spec:
       {{- end }}
+      {{- if .Values.auth.tls.enabled }}
+        - name: init-certs
+          image: busybox
```
As per bitnami's policy, we have to use images created and maintained by us in our charts. Could you modify the used image here to `bitnami/minideb`? We should also include the `securityContext` and `resources` parameters for the initContainer and have this as parameters in `values.yaml`. You could use [this other chart](https://github.com/bitnami/charts/blob/638fdadaa907a36b0304d738cb30aef366cc3255/bitnami/dokuwiki/templates/deployment.yaml#L56) as a reference.

jgallucci32 replied:
Got it. I figured as much just didn't know which image to use. Thanks for the reference.

FraPazGal commented on bitnami/keycloak/templates/statefulset.yaml:
```diff
@@ -59,9 +59,43 @@ spec:
+          - /bin/sh
+          - -c
+          - |-
+            ID="${MY_POD_NAME#${MY_INSTANCE_NAME}-}"
```
I'm not sure `$MY_INSTANCE_NAME` is equivalent to `$fullname := include "keycloak.fullname" .`. If you take a look at the helper's logic, there are cases when these 2 envs are different and this would make `$ID` have a wrong value. I think it would be best to leave this as `$fullname`, WDYT?

jgallucci32 replied:
I'll revert back to original logic to keep it simple. I was trying to remove a dependency on Helm, but if there is a potential for conflict it's not worth the refactor.

FraPazGal reviewed (commented):
Hi @jgallucci32!

It is looking good! There is a couple of changes needed, could you please take a look at my comments?

jgallucci32 commented:
@FraPazGal PR updated with your suggested changes.

FraPazGal commented on bitnami/keycloak/templates/statefulset.yaml:
```diff
@@ -59,9 +59,45 @@ spec:
       {{- end }}
-      {{- if .Values.initContainers }}
+      {{- if or .Values.initContainers .Values.auth.tls.enabled }}
       initContainers: {{- include "common.tplvalues.render" (dict "value" .Values.initContainers "context" $) | nindent 8 }}
       {{- end }}
+      {{- if .Values.auth.tls.enabled }}
+      {{- $fullname := include "keycloak.fullname" . }}
+        - name: init-certs
```
There is an `if` missing here:
```suggestion
      initContainers: 
        {{- if .Values.initContainers }}
        {{- include "common.tplvalues.render" (dict "value" .Values.initContainers "context" $) | nindent 8 }}
        {{- end }}
        {{- if .Values.auth.tls.enabled }}
        {{- $fullname := include "keycloak.fullname" . }}
        - name: init-certs
```

jgallucci32 replied:
Line 62 handles the logical condition for that `if` statement. Please double check, but your suggested change should not be required.

FraPazGal replied:
My bad, I didn't explain myself very well. The `if` I was referring to is `{{- if .Values.initContainers }}`. As it is now, whenever `auth.tls.enabled=true`, the code `{{- include "common.tplvalues.render" (dict "value" .Values.initContainers "context" $) | nindent 8 }}` is going to be executed, which is not something necessary. IMHO, separating this in 2 `ifs` leaves a clearer code and is the default practice in most of our charts.

jgallucci32 replied:
Ah ok I see now. You are correct my logic isn't perfect there. I agree two `if` statements needed to keep the extra include from generating.

FraPazGal commented on bitnami/keycloak/templates/statefulset.yaml:
```diff
@@ -59,9 +59,45 @@ spec:
+          - name: certificates
+            mountPath: /certs
+          - name: shared-certs
+            mountPath: /opt/bitnami/keycloak/certs
+      {{- end }}
```
Missing `end` form the previous if:

```suggestion
            mountPath: /opt/bitnami/keycloak/certs
        {{- end }}
      {{- end }}
```

jgallucci32 replied:
Line `100` ends the if statement in line `65` (I double checked and also counted the if statements). One thing I noticed is that some of the bitnami charts have an extra indent under the `container` and `initContainer` sections. If you look closely enough, the entire section is 2 spaces to the right of what actually gets deployed at runtime. I thought about correcting this but left it alone in case that was done on purpose.

FraPazGal commented on bitnami/keycloak/templates/statefulset.yaml:
```diff
@@ -59,9 +59,45 @@ spec:
+          - |-
+            ID="${MY_POD_NAME#"{{ $fullname }}-"}"
+            if [[ -f "/certs/keycloak.truststore.jks" ]] && [[ -f "/certs/keycloak-${ID}.keystore.jks" ]]; then
+                mkdir -p /opt/bitnami/keycloak/certs
```
As we are mounting `shared-certs` to `/opt/bitnami/keycloak/certs`, we no longer need to create the folder with `mkdir`.

jgallucci32 replied:
Good catch, I will remove that.

FraPazGal commented on bitnami/keycloak/templates/statefulset.yaml:
```diff
@@ -59,9 +59,45 @@ spec:
+          securityContext: {{- omit .Values.containerSecurityContext "enabled" | toYaml | nindent 12 }}
+          {{- end }}
+          command:
+          - /bin/sh
+          - -c
```
Our `bitnami/minideb` image gives some errors if we use `sh`, please change it to `bin/bash`:

```suggestion
          - /bin/bash
          - -ec
```

FraPazGal commented on bitnami/keycloak/values.yaml:
```diff
@@ -92,6 +92,36 @@ auth:
     ## Password to access the truststore when it's password-protected.
     ##
     truststorePassword: ""
+    ## Init container parameters:
```
Please add these new parameters to the README.

jgallucci32 replied:
This has been completed.

FraPazGal reviewed (commented):
Hello @jgallucci32,

I tested this and there are a few issues left I commented on, minor stuff mostly. 

Also, please bump the chart's version to a new minor.

FraPazGal commented on bitnami/keycloak/templates/statefulset.yaml:
```diff
@@ -62,8 +62,46 @@ spec:
+        {{- if .Values.auth.tls.enabled }}
+        {{- $fullname := include "keycloak.fullname" . }}
+        - name: init-certs
+          image: {{ include "keycloak.auth.tls.image" . }}
```
This helper is missing from `_helpers.tlp`. We should include the following:

```
{{/*
Return the proper Auth TLS image name
*/}}
{{- define "keycloak.auth.tls.image" -}}
{{- include "common.images.image" (dict "imageRoot" .Values.auth.tls.image "global" .Values.global) -}}
{{- end -}}
```

jgallucci32 replied:
Thank you for pointing that out. This has been comitted.

FraPazGal commented on bitnami/keycloak/Chart.yaml:
```diff
@@ -26,4 +26,4 @@ name: keycloak
   - https://github.com/bitnami/bitnami-docker-keycloak
   - https://github.com/keycloak/keycloak
-version: 1.4.0
+version: 1.4.1
```
I'm not sure whether to have this as a patch or bump the minor version instead. It is true that this is a fix for an existing functionality, but the changes may be extensive enough for a minor, WDYT?

jgallucci32 replied:
I think minor bump (as-is) since the changes focus solely on the setting `.Values.auth.tls.enabled` being true and getting it to the desired functionality. No new features have been introduced here.

FraPazGal reviewed (commented):
Hello @jgallucci32,

Just a couple of final changes before merging, could you please take a look?

FraPazGal reviewed (approved):
LGTM! Thanks for the great PR!
